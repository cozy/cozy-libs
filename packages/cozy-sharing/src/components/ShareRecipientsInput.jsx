import PropTypes from 'prop-types'
import React from 'react'

import { useQueryAll, isQueryLoading } from 'cozy-client'
import flag from 'cozy-flags'

import ShareAutosuggest from './ShareAutosuggest'
import { getContactsFromGroupId } from '../helpers/contacts'
import {
  buildReachableContactsQuery,
  buildContactGroupsByIdsQuery,
  buildUnreachableContactsWithGroupsQuery
} from '../queries/queries'

/**
 * ShareRecipientsInput is responsible for fetching contacts and groups.
 * We retrieve all the contacts that are reachable and the groups they belong to.
 * In order to display the total number of members in each group, we also retrieve the contacts that are not reachable.
 */
const ShareRecipientsInput = ({
  currentRecipients,
  recipients,
  placeholder,
  onPick,
  onRemove
}) => {
  const reachableContactsQuery = buildReachableContactsQuery()
  const reachableContactsResult = useQueryAll(
    reachableContactsQuery.definition,
    reachableContactsQuery.options
  )

  const unreachableContactsWithGroupsQuery =
    buildUnreachableContactsWithGroupsQuery()
  const unreachableContactsWithGroupsResult = useQueryAll(
    unreachableContactsWithGroupsQuery.definition,
    {
      ...unreachableContactsWithGroupsQuery.options,
      enabled: !!flag('sharing.show-recipient-groups')
    }
  )

  const contactGroupsByIdsQuery = buildContactGroupsByIdsQuery([
    ...new Set(
      (reachableContactsResult.data || []).flatMap(contact => {
        return (contact.relationships?.groups?.data || []).map(
          group => group._id
        )
      })
    )
  ])
  const contactGroupsByIdsResult = useQueryAll(
    contactGroupsByIdsQuery.definition,
    contactGroupsByIdsQuery.options
  )

  const isLoading =
    isQueryLoading(reachableContactsResult) ||
    reachableContactsResult.hasMore ||
    isQueryLoading(contactGroupsByIdsResult) ||
    contactGroupsByIdsResult.hasMore ||
    (flag('sharing.show-recipient-groups') &&
      (isQueryLoading(unreachableContactsWithGroupsResult) ||
        unreachableContactsWithGroupsResult.hasMore))

  const currentRecipientGroups = currentRecipients
    .map(({ id }) => id)
    .filter(Boolean)
  const contactGroups = (contactGroupsByIdsResult.data || [])
    .filter(
      ({ _id }) =>
        !flag('sharing.show-recipient-groups') ||
        !currentRecipientGroups.includes(_id)
    )
    .map(contactGroup => {
      const reachableMembers = getContactsFromGroupId(
        reachableContactsResult.data,
        contactGroup.id
      ).map(contact => ({
        ...contact,
        isReachable: true
      }))

      if (flag('sharing.show-recipient-groups') !== true) {
        return {
          ...contactGroup,
          members: reachableMembers
        }
      }

      const unreachableMembers = getContactsFromGroupId(
        unreachableContactsWithGroupsResult.data,
        contactGroup.id
      ).map(contact => ({
        ...contact,
        isReachable: false
      }))

      return {
        ...contactGroup,
        members: [...reachableMembers, ...unreachableMembers]
      }
    })

  const contactsAndGroups = [
    ...(reachableContactsResult.data || []),
    ...contactGroups
  ]

  return (
    <ShareAutosuggest
      loading={isLoading}
      contactsAndGroups={contactsAndGroups}
      recipients={recipients}
      onPick={onPick}
      onRemove={onRemove}
      placeholder={placeholder}
    />
  )
}

ShareRecipientsInput.propTypes = {
  currentRecipients: PropTypes.array,
  recipients: PropTypes.array,
  placeholder: PropTypes.string,
  onPick: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired
}

ShareRecipientsInput.defaultProps = {
  currentRecipients: [],
  recipients: []
}

export default ShareRecipientsInput
