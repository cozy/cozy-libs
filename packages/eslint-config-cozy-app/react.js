'use strict'

const basics = require('./basics')

module.exports = {
  plugins: basics.plugins.concat(['react-hooks']),
  extends: basics.extends.concat(['plugin:react/recommended']),
  parser: basics.parser,
  parserOptions: { ecmaFeatures: { jsx: true } },
  env: basics.env,
  settings: { react: { version: 'latest' } },
  rules: Object.assign({}, basics.rules, {
    'react/prop-types': 'off',
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'warn'
  })
}
