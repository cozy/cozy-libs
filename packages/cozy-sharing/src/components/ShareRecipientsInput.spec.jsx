import { render, screen } from '@testing-library/react'
import React from 'react'

import { createMockClient } from 'cozy-client'

import ShareRecipientsInput from './ShareRecipientsInput'
import AppLike from './SharingBanner/test/AppLike'

jest.mock('./ShareAutosuggest', () => ({
  __esModule: true,
  default: ({ contactsAndGroups }) => {
    return (
      <ul>
        {contactsAndGroups.map(contactOrGroup => {
          if (contactOrGroup._type === 'io.cozy.contacts') {
            return (
              <li key={contactOrGroup.id}>
                {contactOrGroup.name.givenName} {contactOrGroup.name.familyName}
              </li>
            )
          } else {
            return <li key={contactOrGroup.id}>{contactOrGroup.name}</li>
          }
        })}
      </ul>
    )
  }
}))

describe('ShareRecipientsInput component', () => {
  const setup = ({ contacts, contactGroups }) => {
    const mockClient = createMockClient({
      queries: {
        'io.cozy.contacts': {
          doctype: 'io.cozy.contacts',
          data: contacts,
          meta: { count: 4 }
        },
        'io.cozy.contacts.groups': {
          doctype: 'io.cozy.contacts.groups',
          data: contactGroups,
          hasMore: false
        }
      }
    })
    render(
      <AppLike client={mockClient}>
        <ShareRecipientsInput
          onPick={() => {}}
          onRemove={() => {}}
          placeholder="Enter recipients here"
        />
      </AppLike>
    )
  }

  it('should include groups and contacts when are loaded', async () => {
    const contacts = [
      {
        id: 'df563cc4-6440',
        _id: 'df563cc4-6440',
        _type: 'io.cozy.contacts',
        name: {
          givenName: 'Michale',
          familyName: 'Russel'
        }
      },
      {
        id: '5a3b4ccf-c257',
        _id: '5a3b4ccf-c257',
        _type: 'io.cozy.contacts',
        name: {
          givenName: 'Teagan',
          familyName: 'Wolf'
        },
        relationships: {
          groups: {
            data: [
              {
                _id: 'fe86af20-c6c5',
                _type: 'io.cozy.contacts.groups'
              }
            ]
          }
        }
      }
    ]

    const contactGroups = [
      {
        id: 'fe86af20-c6c5',
        name: "The Night's Watch",
        _id: 'fe86af20-c6c5',
        _type: 'io.cozy.contacts.groups'
      }
    ]

    setup({ contacts, contactGroups })

    expect(screen.getByText('Michale Russel')).toBeInTheDocument()
    expect(screen.getByText('Teagan Wolf')).toBeInTheDocument()
    expect(screen.getByText("The Night's Watch")).toBeInTheDocument()
  })

  it('should include groups only if they have more than one member', async () => {
    const contacts = [
      {
        id: 'df563cc4-6440',
        _id: 'df563cc4-6440',
        _type: 'io.cozy.contacts',
        name: {
          givenName: 'Michale',
          familyName: 'Russel'
        }
      },
      {
        id: '5a3b4ccf-c257',
        _id: '5a3b4ccf-c257',
        _type: 'io.cozy.contacts',
        name: {
          givenName: 'Teagan',
          familyName: 'Wolf'
        }
      }
    ]

    const contactGroups = [
      {
        id: 'fe86af20-c6c5',
        name: "The Night's Watch",
        _id: 'fe86af20-c6c5',
        _type: 'io.cozy.contacts.groups'
      }
    ]

    setup({ contacts, contactGroups })

    expect(screen.getByText('Michale Russel')).toBeInTheDocument()
    expect(screen.getByText('Teagan Wolf')).toBeInTheDocument()
    expect(screen.queryByText("The Night's Watch")).toBeNull()
  })
})
