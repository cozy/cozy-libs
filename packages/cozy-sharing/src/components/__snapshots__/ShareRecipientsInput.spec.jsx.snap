// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareRecipientsInput component should include groups only if all contacts are loaded 1`] = `
<ShareAutocomplete
  contactsAndGroups={
    Array [
      Object {
        "_id": "df563cc4-6440",
        "_type": "io.cozy.contacts",
        "id": "df563cc4-6440",
        "name": Object {
          "familyName": "Russel",
          "givenName": "Michale",
        },
      },
      Object {
        "_id": "5a3b4ccf-c257",
        "_type": "io.cozy.contacts",
        "id": "5a3b4ccf-c257",
        "name": Object {
          "familyName": "Wolf",
          "givenName": "Teagan",
        },
      },
    ]
  }
  loading={false}
  onFocus={[Function]}
  onPick={[MockFunction onPick]}
  onRemove={[MockFunction onRemove]}
  placeholder="Enter recipients here"
  recipients={
    Array [
      Object {
        "id": "5a3b4ccf-c257",
        "name": Object {
          "familyName": "Wolf",
          "givenName": "Teagan",
        },
      },
    ]
  }
/>
`;

exports[`ShareRecipientsInput component should match snapshot 1`] = `
<ShareAutocomplete
  contactsAndGroups={
    Array [
      Object {
        "_id": "df563cc4-6440",
        "_type": "io.cozy.contacts",
        "id": "df563cc4-6440",
        "name": Object {
          "familyName": "Russel",
          "givenName": "Michale",
        },
      },
      Object {
        "_id": "5a3b4ccf-c257",
        "_type": "io.cozy.contacts",
        "id": "5a3b4ccf-c257",
        "name": Object {
          "familyName": "Wolf",
          "givenName": "Teagan",
        },
      },
      Object {
        "_id": "fe86af20-c6c5",
        "_type": "io.cozy.contacts.groups",
        "id": "fe86af20-c6c5",
        "name": "The Night's Watch",
      },
      Object {
        "_id": "3d8193ab-2ce4",
        "_type": "io.cozy.contacts.groups",
        "id": "3d8193ab-2ce4",
        "name": "The North",
      },
    ]
  }
  loading={false}
  onFocus={[Function]}
  onPick={[MockFunction onPick]}
  onRemove={[MockFunction onRemove]}
  placeholder="Enter recipients here"
  recipients={
    Array [
      Object {
        "id": "5a3b4ccf-c257",
        "name": Object {
          "familyName": "Wolf",
          "givenName": "Teagan",
        },
      },
    ]
  }
/>
`;
